rkt/trust.go

```diff
@@ -0,0 +1,230 @@
+// limitations under the License.
+
+//+build linux
+
+// implements https://github.com/coreos/rocket/issues/367
+
+package main
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+
+	"github.com/coreos/rocket/Godeps/_workspace/src/golang.org/x/crypto/openpgp"
+	"github.com/coreos/rocket/pkg/keystore"
+)
+
+var (
+	cmdTrust = &Command{
+		Name:    cmdTrustName,
+		Summary: "Trust a key for image verification",
+		Usage:   "[--prefix PREFIX] [--do-not-ask-confirmation] PUBKEY",
+		Run:     runTrust,
+	}
+	flagPrefix string
+	flagAccept bool
+)
+
+const (
+	cmdTrustName = "trust"
+)
+
+func init() {
+	commands = append(commands, cmdTrust)
+	cmdTrust.Flags.StringVar(&flagPrefix, "prefix", "", "prefix to limit trust to")
+	cmdTrust.Flags.BoolVar(&flagAccept, "do-not-ask-confirmation", false, "suppress interactive confirmation of key addition")
+}
+
+func runTrust(args []string) (exit int) {
+	if len(args) != 1 {
+		printCommandUsageByName(cmdTrustName)
+		return 1
+	}
+
+	pkl, err := getPubKeyLocation(flagPrefix, args)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Error determining key location: %v\n", err)
+		return 1
+	}
+
+	pk, err := getPubKey(pkl)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Error accessing key: %v\n", err)
+		return 1
+	}
+	defer pk.Close()
+
+	accepted, err := reviewKey(pk, flagAccept)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Error reviewing key: %v\n", err)
+	}
+
+	if !accepted {
+		fmt.Fprintf(os.Stderr, "Not trusting %q.\n", pkl)
+		return 0
+	}
+
+	fmt.Printf("Trusting %q for prefix %q.\n", pkl, flagPrefix)
+
+	if err := addPubKey(flagPrefix, pk); err != nil {
+		fmt.Fprintf(os.Stderr, "Error adding key: %v\n", err)
+		return 1
+	}
+
+	return 0
+}
+
+// getPubKeyLocation either returns the location supplied in argv or discovers one @ prefix
+func getPubKeyLocation(prefix string, args []string) (string, error) {
+	if len(args) == 1 {
+		return args[0], nil
+	}
+
+	if len(args) > 1 {
+		return "", fmt.Errorf("too many arguments")
+	}
+
+	if prefix == "" {
+		return "", fmt.Errorf("a key or --prefix required")
+	}
+
+	// TODO(vc): discover location from --prefix and return it here?
+
+	return "", nil
+}
+
+// getPubKey retrieves a public key (if remote), and verifies it's a gpg key
+func getPubKey(location string) (*os.File, error) {
+	u, err := url.Parse(location)
+	if err != nil {
+		return nil, err
+	}
+
+	switch u.Scheme {
+	case "":
+		return os.Open(location)
+	case "http":
+		fallthrough
+	case "https":
+		return downloadKey(u.String())
+	default:
+	}
+	return nil, fmt.Errorf("only http and https urls supported")
+}
+
+// downloadKey retrieves the file, storing it in a deleted tempfile
+func downloadKey(url string) (*os.File, error) {
+	tf, err := ioutil.TempFile("", "")
+	if err != nil {
+		return nil, fmt.Errorf("error creating tempfile: %v", err)
+	}
+	os.Remove(tf.Name()) // no need to keep the tempfile around
```

Comment:
eh?  There are some missing tf.Close()'s, but tf needs to be returned open in the success case.